modules/standard/Memory/Initialization.chpl

```diff
@@ -0,0 +1,290 @@
+    import HaltWrappers.boundsCheckHalt;
+
+    if !a.domain.contains(idx) then
+      boundsCheckHalt('Cannot move-initialize array because its domain ' +
+                      'does not contain: ' + indexName);
+  }
+
+  // Call to 'indexOrder' assumes 'idx' is a valid index.
+  private inline proc _haltBadElementRange(a, idx, numElements: int) {
+    import HaltWrappers.boundsCheckHalt;
+
+    if numElements > a.size then
+      boundsCheckHalt('Cannot move-initialize array because number of ' +
+                      'elements to copy exceeds array size');
+
+    if numElements <= 0 then
+      boundsCheckHalt('Cannot move-initialize array because number of ' +
+                      'elements to copy is <= 0');
+
+    const order = a.domain.indexOrder(idx);
+    const hi = order + numElements;
+
+    if hi > a.size then
+      boundsCheckHalt('Cannot move-initialize array because one or ' +
+                      'more indices are outside its domain');
+  }
+
+  private inline proc _errorNonCopyableType(type t) {
+    if !isCopyableType(t) then
+      compilerError('Cannot move-initialize array with non-copyable ',
+                    'element type: ', t:string, 2);
+  }
+
+  // TODO: Relax this restriction in the future.
+  private inline proc _errorNot1DRectangularArray(a) {
+    if !a.isDefaultRectangular() || a.rank > 1 then
+      compilerError('Can only move-initialize one-dimensional ',
+                    'rectangular arrays', 2);
+  }
+
+  /*
+    Move-initialize a group of array elements from a different group of
+    elements in the same array.
+
+    :arg a: An array to move-initialize
+    :type a: `[]`
+
+    :arg dstStartIndex: Destination index of elements to move-initialize
+    :type dstStartIndex: `a.idxType`
+
+    :arg srcStartIndex: Source index of elements to move-initialize from
+    :type srcStartIndex: `a.idxType`
+
+    :arg numElements: The number of elements to move-initialize
+    :type numElements: int
+  */
+  proc moveInitializeArrayElements(ref a: [?d], dstStartIndex: a.idxType,
+                                   srcStartIndex: a.idxType,
+                                   numElements: int) {
+    _errorNonCopyableType(a.eltType);
+    _errorNot1DRectangularArray(a);
+
+    if boundsChecking { 
+      _haltBadElementRange(a, dstStartIndex, numElements);
+      _haltBadIndex(a, dstStartIndex, 'dstStartIndex');
+      _haltBadIndex(a, srcStartIndex, 'srcStartIndex');
+    }
+
+    if dstStartIndex == srcStartIndex then
+      return;
+
+    const d = a.domain;
+
+    const dstLo = d.indexOrder(dstStartIndex);
+    const srcLo = d.indexOrder(srcStartIndex);
+
+    forall i in 0..<numElements {
+      const dstIdx = d.orderToIndex(dstLo + i);
+      const srcIdx = d.orderToIndex(srcLo + i);
+      ref dst = a[dstIdx];
+      ref src = a[srcIdx];
+      moveInitialize(dst, src);
+    }
+  }
+
+  /*
+    Move-initialize a group of array elements from a different group of
+    elements in another array.
```

Comment:
Remove "different"
Add, as above: This is equivalent to and potentially better-performing than a sequence of individual calls to `moveInitialize().`